raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java

```diff
@@ -2132,15 +2132,32 @@ private long pollLeader(long currentTimeMs) {
             currentTimeMs
         );
 
-        long timeUntilSend = maybeSendRequests(
-            currentTimeMs,
-            partitionState
-                .lastVoterSet()
-                .voterNodes(state.nonAcknowledgingVoters().stream(), channel.listenerName()),
-            this::buildBeginQuorumEpochRequest
-        );
+        long timeUntilNextBeginQuorumSend = state.timeUntilBeginQuorumEpochTimerExpires(currentTimeMs);
+
+        if (timeUntilNextBeginQuorumSend == 0) {
+            timeUntilNextBeginQuorumSend = maybeSendRequests(
+                currentTimeMs,
+                partitionState
+                    .lastVoterSet()
+                    .voterNodes(channel.listenerName()),
+                this::buildBeginQuorumEpochRequest
+            );
+            state.resetBeginQuorumEpochTimer(currentTimeMs);
+            logger.trace("Attempted to send BeginQuorumEpochRequest as heartbeat to all voters. " +
+                "Request can be retried in {} ms", timeUntilNextBeginQuorumSend);
+        } else if (!state.nonAcknowledgingVoters().isEmpty()) {
+            timeUntilNextBeginQuorumSend = maybeSendRequests(
+                currentTimeMs,
+                partitionState
+                    .lastVoterSet()
+                    .voterNodes(state.nonAcknowledgingVoters().stream(), channel.listenerName()),
+                this::buildBeginQuorumEpochRequest
+            );
+            logger.trace("Attempted to send BeginQuorumEpochRequest to non-acknowledging voters: {}. " +
+                "Request can be retried in {} ms", state.nonAcknowledgingVoters(), timeUntilNextBeginQuorumSend);
+        }
```

Comment:
How about moving this algorithm to a private method? So that the changes to`pollLeader` become `maybeSendBeginQuorumEpochRequests(currentTimeMs);`.